src/python/pants/core/goals/deploy.py

```diff
@@ -0,0 +1,203 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+from abc import ABCMeta
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.core.goals.package import PackageFieldSet
+from pants.core.goals.publish import PublishFieldSet, PublishProcesses, PublishProcessesRequest
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.process import InteractiveProcess, InteractiveProcessResult
+from pants.engine.rules import Effect, Get, MultiGet, collect_rules, goal_rule, rule_helper
+from pants.engine.target import (
+    FieldSet,
+    FieldSetsPerTarget,
+    FieldSetsPerTargetRequest,
+    NoApplicableTargetsBehavior,
+    Target,
+    TargetRootsToFieldSets,
+    TargetRootsToFieldSetsRequest,
+)
+from pants.engine.unions import union
+from pants.option.option_types import ArgsListOption
+
+logger = logging.getLogger(__name__)
+
+
+@union
+@dataclass(frozen=True)
+class DeployFieldSet(FieldSet, metaclass=ABCMeta):
+    """The FieldSet type for the `deploy` goal."""
+
+
+@dataclass(frozen=True)
+class DeployProcess:
```

Comment:
Maybe add a doc string?

Some questions to answer in a doc string:
- How is this type requested? I assume the engine is asked for `DeployProcess` given a `DeployFieldSet`?
- How does a `publish_dependency` differ from a dependency on a target matching a `DeployFieldSet`?